src/app/clusters/icd-management-server/icd-management-server.cpp

```diff
@@ -137,76 +149,213 @@ CHIP_ERROR IcdManagementAttributeAccess::ReadRegisteredClients(EndpointId endpoi
+    return encoder.Encode(mICDData->GetICDCounter());
 }
 
 CHIP_ERROR IcdManagementAttributeAccess::ReadClientsSupportedPerFabric(EndpointId endpoint, AttributeValueEncoder & encoder)
 {
-    return encoder.Encode(ICDManagementServer::GetInstance().GetClientsSupportedPerFabric());
+    return encoder.Encode(mICDData->GetClientsSupportedPerFabric());
 }
 
-CHIP_ERROR CheckAdmin(chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath, bool & is_admin)
+/**
+ * @brief Implementation of Fabric Delegate for ICD Management cluster
+ */
+class IcdManagementFabricDelegate : public FabricTable::Delegate
 {
-    RequestPath requestPath{ .cluster = commandPath.mClusterId, .endpoint = commandPath.mEndpointId };
-    CHIP_ERROR err = GetAccessControl().Check(commandObj->GetSubjectDescriptor(), requestPath, Privilege::kAdminister);
-    if (CHIP_NO_ERROR == err)
-    {
-        is_admin = true;
-    }
-    else if (CHIP_ERROR_ACCESS_DENIED == err)
+public:
+    void Init(PersistentStorageDelegate & storage, Crypto::SessionKeystore * sessionKeystore, ICDData & icdData)
     {
-        is_admin = false;
-        err      = CHIP_NO_ERROR;
+        mStorage         = &storage;
+        mSessionKeystore = sessionKeystore;
+        mICDData         = &icdData;
     }
-    return err;
-}
 
-class IcdManagementFabricDelegate : public chip::FabricTable::Delegate
-{
     void OnFabricRemoved(const FabricTable & fabricTable, FabricIndex fabricIndex) override
     {
-        uint16_t supported_clients = ICDManagementServer::GetInstance().GetClientsSupportedPerFabric();
-        ICDMonitoringTable table(Server::GetInstance().GetPersistentStorage(), fabricIndex, supported_clients,
-                                 Server::GetInstance().GetSessionKeystore());
+        uint16_t supported_clients = mICDData->GetClientsSupportedPerFabric();
+        ICDMonitoringTable table(*mStorage, fabricIndex, supported_clients, mSessionKeystore);
         table.RemoveAll();
-        ICDManagementServer::GetInstance().TriggerICDMTableUpdatedEvent();
+        app::ICDNotifier::GetInstance().BroadcastICDManagementEvent(app::ICDListener::ICDManagementEvents::kTableUpdated);
     }
+
+private:
+    PersistentStorageDelegate * mStorage       = nullptr;
+    Crypto::SessionKeystore * mSessionKeystore = nullptr;
+    ICDData * mICDData                         = nullptr;
 };
 
 IcdManagementFabricDelegate gFabricDelegate;
 IcdManagementAttributeAccess gAttribute;
 
 } // namespace
 
-void emberAfIcdManagementClusterInitCallback()
+/*
+ * ICD Management Implementation
+ */
+
+PersistentStorageDelegate * ICDManagementServer::mStorage           = nullptr;
+Crypto::SymmetricKeystore * ICDManagementServer::mSymmetricKeystore = nullptr;
+ICDData * ICDManagementServer::mICDData                             = nullptr;
+
+Status ICDManagementServer::RegisterClient(FabricIndex fabric_index, NodeId node_id, uint64_t monitored_subject, ByteSpan key,
```

Comment:
At some point we may want to create some "parameter" class as number of arguments to methods increase. In this case 7 starts being quite large and not very obvious what they mean as one is a bool (always annoying) and 2 are pure numeric.

Reply:
Agreed; next step in the refactor is to use the commandObj, commandData and commandPath as inputs to these functions and process them inside the function. 

Initial clean up was getting a bit big for this to be included (imo)

Reply:
https://github.com/project-chip/connectedhomeip/issues/30631